superset/views/core.py

```diff
@@ -2503,6 +2504,61 @@ def stop_query(self):
             pass
         return self.json_response('OK')
 
+    @has_access_api
+    @expose('/validate_sql_json/', methods=['POST', 'GET'])
+    @log_this
+    def validate_sql_json(self):
+        """Validates that arbitrary sql is acceptable for the given database.
+        Returns a list of error/warning annotations as json.
+        """
+        sql = request.form.get('sql')
+        database_id = request.form.get('database_id')
+        schema = request.form.get('schema') or None
+        template_params = json.loads(
+            request.form.get('templateParams') or '{}')
+
+        if len(template_params) > 0:
+            # TODO: factor the Database object out of template rendering
+            #       or provide it as mydb so we can render template params
+            #       without having to also persist a Query ORM object.
+            return json_error_response(
+                'SQL validation does not support template parameters')
+
+        session = db.session()
+        mydb = session.query(models.Database).filter_by(id=database_id).first()
+        if not mydb:
+            json_error_response(
+                'Database with id {} is missing.'.format(database_id))
+
+        spec = mydb.db_engine_spec
+        if not spec.engine in SQL_VALIDATORS_BY_ENGINE:
+            return json_error_response(
+                'no SQL validator is configured for {}'.format(spec.engine))
+        validator = SQL_VALIDATORS_BY_ENGINE[spec.engine]
+
+        try:
+            timeout = config.get('SQLLAB_VALIDATION_TIMEOUT')
+            timeout_msg = (
+                f'The query exceeded the {timeout} seconds timeout.')
+            with utils.timeout(seconds=timeout,
+                                error_message=timeout_msg):
+                errors = validator.validate(sql, schema, mydb)
+            payload = json.dumps(
+                [err.to_dict() for err in errors],
+                default=utils.pessimistic_json_iso_dttm_ser,
+                ignore_nan=True,
+                encoding=None,
+            )
+            return json_success(payload)
+        except Exception as e:
+            logging.exception(e)
+            msg = _(
+                'Failed to validate your SQL query text. Please check that '
+                f'you have configured the {validator.name} validator '
+                'correctly and that any services it depends on are up. '
+                f'Exception: {e}')
+            return json_error_response(f'{msg}')
```

Comment:
I would suggest a 400 (bad request) here... if the user sends malformed, unparseable SQL then this is on them, and does not reflect on the health of the service.

Reply:
This condition's raised whenever the validator excepts out for any reason we didn't predict — it could be a bad request, but the most likely reason to end up on this codepath is that the validator itself errored out. For presto that probably means the backing DB 5xx'd us.

On the client side I'm showing this as a warning box in place of the control that normally shows you what the error text is, saying `The server failed to validate your query using ${validator_name}` — just to let you know "hey, the reason you're not seeing errors highlighted isn't that your query is good, it's just that we don't know if it's good or bad"

Reply:
I'm actually pretty concerned about this... this would allow any user to just spam the backend with bad queries and wake up an on-call at 2am.  Can the validator itself make the distinction between an upstream 500 response from the db and a "couldn't parse" problem?

Reply:
I'm not sure what that distinction would look like — here's the mental model I'm using:

The `validator.validate()` interface should never throw, unless it broke. Given that, anything that throws out from it is legitimately a sever error, and should contribute to paging the oncall — if a user is able to form requests that cause the validator to 5xx, that's legitimate breakage.

Phrased another way, if it shows up as a 5xx when someone tries to run a synchronous query and the backing db is toast, shouldn't validations do the same?

Reply:
OK after our in-person chat I'm sold! Thanks 😆 Def add some explanatory comment to that block on `presto_db.py` line 91.

Reply:
Added! Give that a once-over and let me know if that helps to clarify how that part works, since it's definitely not obvious from the code itself.

Reply:
yeah that is much more clear, thanks a ton.